### PR TITLE
Adding an exponential back-off logic to rename files

### PR DESCRIPTION
**Pull request by gkc2104**
## Not ready for Merge

Creating a PR to get some feedback as to how this needs to be implemented, and I went ahead and made a very basic implementation to start off. 

TODO:
 - make `retryThreshold ` and `retryGranularity` configurable.
 - add test cases

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 1]
```diff
@@ -454,7 +454,25 @@ public long push() throws IOException
...
+        }
+        catch (InterruptedException ex) {
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
+
+    if (retryAttempts == retryThreshold) {
       throw new IOException(
           String.format(
```
Comment (jisookim0513) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Is it possible to record the reason the renaming failed?

Comment (gkc2104) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: I looked at `renameIndexFiles()` and `FileSystem.rename()` they seem to capture all fail cases I can think of, I'm not sure what the exact reason is in this case

File: indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java [hunk 2]
```diff
@@ -454,7 +454,25 @@ public long push() throws IOException
         .withSize(size.get())
         .withBinaryVersion(SegmentUtils.getVersionFromDir(mergedBase));
 
-    if (!renameIndexFiles(outputFS, tmpPath, finalIndexZipFilePath)) {
+    int retryAttempts = 0;
+    int retryThreshold = 5;
+    int retryGranularity = 1000;
+
+    while (retryAttempts != retryThreshold) {
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Please use RetryUtils

Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java]: Note that RetryUtils does things like retry jitter io.druid.java.util.common.RetryUtils#awaitNextRetry

**Comment by drcrallen** [comment]
Discussed with @gkc2104 , the errors are logged under the RetryUtils class rather than the one that was expected. It was observed that there were some responsiveness problems to S3 which were reported in RetryUtils then succeeded on the next attempt.

**Pull request closed without merging**